### PR TITLE
Add batch job as a cloudwatch_event_target

### LINKED ISSUE #4023: Add AWS Batch as a cloudwatch_event_target
This is a feature request.

AWS has recently added the ability to use an AWS Batch resource as the target of a cloudwatch event: https://aws.amazon.com/about-aws/whats-new/2018/03/amazon-cloudwatch-events-now-supports-aws-batch-as-an-event-target/

This is confirmed by the AWS documentation: https://docs.aws.amazon.com/AmazonCloudWatchEvents/latest/APIReference/API_Target.html as it now accepts a BatchParameters object

In Terraform we would need to add to the `cloudwatch_event_target` resource a `BatchParameters` corresponding field - maybe something like `batch_target`

### PR DESCRIPTION
**Pull request by gheine**
Fixes #4023

### CODE REVIEW

File: aws/resource_aws_cloudwatch_event_target.go [hunk 1]
```diff
@@ -102,6 +102,34 @@ func resourceAwsCloudWatchEventTarget() *schema.Resource {
...
+							Required: true,
+						},
+						"array_size": {
+							Type:         schema.TypeInt,
+							Optional:     true,
+							ValidateFunc: validation.IntBetween(2, 10000),
+						},
+						"job_attempts": {
+							Type:         schema.TypeInt,
+							Required:     true,
```
[the remark this answers is not in the document]
Comment (gheine) [reply aws/resource_aws_cloudwatch_event_target.go]: Dang! good spot!

File: aws/resource_aws_cloudwatch_event_target_test.go [hunk 2]
```diff
@@ -473,6 +493,155 @@ EOF
...
+  schedule_expression = "rate(5 minutes)"
+}
+
+resource "aws_cloudwatch_event_target" "test" {
+  arn = "${aws_batch_job_queue.batch_job_queue.arn}"
+  rule = "${aws_cloudwatch_event_rule.cloudwatch_event_rule.id}"
+  role_arn = "${aws_iam_role.event_iam_role.arn}"
+
+  batch_target {
+    job_definition = "${aws_batch_job_definition.batch_job_definition}"
```
Comment (bflad) [inline aws/resource_aws_cloudwatch_event_target_test.go]:
```
=== RUN   TestAccAWSCloudWatchEventTarget_batch
--- FAIL: TestAccAWSCloudWatchEventTarget_batch (1.14s)
    testing.go:518: Step 0 error: Error loading configuration: Error loading /opt/teamcity-agent/temp/buildTmp/tf-test953762704/main.tf: Error reading config for aws_cloudwatch_event_target[test]: aws_batch_job_definition.batch_job_definition: resource variables must be three parts: TYPE.NAME.ATTR in:
        
        ${aws_batch_job_definition.batch_job_definition}
```

File: aws/resource_aws_cloudwatch_event_target_test.go [hunk 3]
```diff
@@ -473,6 +493,155 @@ EOF
...
+  service_role = "${aws_iam_role.batch_iam_role.arn}"
+  type = "MANAGED"
+  depends_on = ["aws_iam_role_policy_attachment.aws_batch_service_role"]
+}
+
+resource "aws_batch_job_queue" "batch_job_queue" {
+  name = "%[1]s"
+  state = "ENABLED"
+  priority = 1
+  compute_environments = ["${aws_batch_compute_environment.batch_compute_environment}"]
```
Comment (bflad) [inline aws/resource_aws_cloudwatch_event_target_test.go]: Missing attribute here as well.

File: aws/resource_aws_cloudwatch_event_target.go [hunk 4]
```diff
@@ -344,6 +381,27 @@ func expandAwsCloudWatchEventTargetEcsParameters(config []interface{}) *events.E
...
+	for _, c := range config {
+		param := c.(map[string]interface{})
+		batchParameters.JobDefinition = aws.String(param["job_definition"].(string))
+		batchParameters.JobName = aws.String(param["job_name"].(string))
+		if v := param["array_size"]; v != nil {
+			arrayProperties := &events.BatchArrayProperties{}
+			arrayProperties.Size = aws.Int64(int64(v.(int)))
+			batchParameters.ArrayProperties = arrayProperties
+		}
+		if v := param["job_attempts"]; v != nil {
```
[the remark this answers is not in the document]
Comment (gheine) [reply aws/resource_aws_cloudwatch_event_target.go]: Fixed

**Comment by bflad** [comment]
This has been released in [version 1.16.0 of the AWS provider](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md#1160-april-25-2018). Please see the [Terraform documentation on provider versioning](https://www.terraform.io/docs/configuration/providers.html#provider-versions) or reach out if you need any assistance upgrading.

**Pull request merged**